Wrapping exception

thomaslevesque opened the pull request:
Fixes #836

thomaslevesque commented on src/FakeItEasy/Core/WrappedObjectRule.cs:
```diff
@@ -50,8 +58,32 @@ public void Apply(IInterceptedFakeObjectCall fakeObjectCall)
+            throw e;
+#else
+            ExceptionDispatchInfo.Capture(e).Throw();
+#endif
```
I marked the PR as [WIP] because of this ugly hack. `ExceptionDispatchInfo` is only available since .NET 4.5, so I had to resort to this solution (found [here](http://stackoverflow.com/a/2085377/98713)). Another approach involves calling an internal method (`InternalPreserveStacktrace`) via reflection... not sure which is better.

blairconrad replied:
I thought I'd made this comment earlier. Sorry if it's a duplicate. Consider looking at [FakeManager.TryPreserveStackTrace](https://github.com/FakeItEasy/FakeItEasy/blob/master/src/FakeItEasy/Core/FakeManager.EventRule.cs#L72). Not because it's necessarily better, but because it does (a subset of) what you're doing here, and we should probably have only the one way.

thomaslevesque replied:
Ah, I thought we must already have something like that, but didn't find it. Thanks!
However, it works on the full .NET Framework, but probably not on .NET Core... We must have missed it during the port to .NET Core because of the empty catch. I'll extract it to a separate class and make it work for .NET Core.

thomaslevesque replied:
Actually, the `InternalPreserveStackTrace` trick _does_ work on .NET Core. But it's probably still better to use `ExceptionDispatchInfo`, since it's public.

blairconrad commented on tests/FakeItEasy.Specs/WrappingFakeSpecs.cs:
```diff
@@ -0,0 +1,169 @@
+            "And a fake wrapping this object"
+                .x(() => wrapper = A.Fake<IFoo>(o => o.Wrapping(realObject)));
+
+            "When a method with out and ref parameters is called is called on the wrapper"
```
"is called is called"

thomaslevesque replied:
oops!

blairconrad commented on tests/FakeItEasy.Specs/WrappingFakeSpecs.cs:
```diff
@@ -0,0 +1,169 @@
+            public void OutAndRefMethod(ref int refParameter, out int outParameter)
+            {
+                this.OutAndRefMethodCalled = true;
+                refParameter += 1;
```
Why a `+=` here, instead of a `=`? It works, but in theory could make tests more brittle, or require more logic in them.

thomaslevesque replied:
Because it's a ref param, so I wanted to ensure that we were passing the original value. I know it's a bit strange, especially since I don't initialize the parameter before the call, so it's always 0 (I never know where to initialize things in XBehave specs). If you prefer I can change it to just `refParameter = 1`.

blairconrad replied:
That's okay, I'm content.

adamralph replied:
> I never know where to initialize things in XBehave specs

If the initialisation is a first class part of the scenario, then it should be mentioned in the natural language as one of the steps, e.g.

```gherkin
Given a real object
  And a fake wrapping the object
  And a value of zero
 When the value is passed by reference to a method which increments it
...
```

If it's irrelevant to the natural language, you can do it anywhere, e.g. at the last moment before the value is used.

thomaslevesque replied:
In this case, no, it's not really a first class part of the scenario. It's just the initial value of a `ref` parameter. Is it okay to do it between steps?

adamralph replied:
You shouldn't really do anything outside steps. The entire body of the test should really be inside the step delegates. Doing some stuff outside them will work, other stuff won't, so I think it's much better to keep everything inside them, e.g.

```C#
"When a method with out and ref parameters is called on the wrapper"
    .x(() =>
    {
        refParameter = 0;
        wrapper.OutAndRefMethod(ref refParameter, out outParameter);
    });
```

However, if you know that the wrapped object increments the ref param, then aren't the initial and final values fundamental to the scenario?

adamralph replied:
I.e. ignoring `out` for a moment (perhaps it's better to split the scenario?) you could write the scenario like so:

```gherkin
Given an object with a method that increments an integer passed by reference
  And a fake which wraps the object
  And an integer
 When I pass the integer to the fake method
 Then the integer is incremented
```

Or being less specific:

```gherkin
Given an object with a method that assigns an argument passed by reference
  And a fake which wraps the object
  And another object
 When I pass the second object to the fake method
 Then the object is reassigned
```

adamralph replied:
I guess the point I'm trying to make is that the natural language should adequately describe the scenario on it's own. I.e. you should be able to take the code away and still understand what the scenario is and the steps it describes.

thomaslevesque replied:
@adamralph thanks for the explanation.

> However, if you know that the wrapped object increments the ref param, then aren't the initial and final values fundamental to the scenario?

Yes, you're right. I'll update the specs to initialize it explicitly.

thomaslevesque commented:
OK, I'm clearly missing something about netstd/project.json

I tried to add a conditional compilation symbol in project.json:

```
"compilationOptions": {
    "define": [ "FEATURE_EXCEPTION_DISPATCH_INFO" ]
  },
```

but it has no effect at all. I had to put it in FakeItEasy.netstd.csproj instead. So what's the point of having project.json ? Is it only for dependencies?

thomaslevesque commented:
OK, I made the requested changes, except for the `+=`; it kinda bothers me to overwrite a ref param without even looking at its value, and ReSharper yells at me. If you think it's important, I'll change it.

I also added spec steps to check that the exception stack trace is preserved.

blairconrad commented:
@thomaslevesque, thanks for at least thinking about the `+=`. I'm satisfied with your reasoning.
Also, thanks for including the `FakeManager` in your change.

blairconrad commented:
> So what's the point of having project.json

Darned if I know. I'm seeing them have more effect in an `xproj`-based project. And they seem to be _the place_ for specifying your multi-targeting. But, yeah, the whole thing is magic to me still. I'm flailing about, hoping things work, and only rarely do I understand why they do when they do.

blairconrad commented on src/FakeItEasy/ExceptionExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+            ExceptionDispatchInfo.Capture(exception).Throw();
+        }
+#else
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Appropriate in try method.")]
```
Nit: it's no longer a "Try*" method.

thomaslevesque replied:
True, I'll amend

blairconrad commented:
@thomaslevesque, it all looks good to me. Thanks!
Pausing for a bit in case any other comments come in.

thomaslevesque commented:
- fixed the justification in the `[SuppressMessage]` attribute
- cached a delegate for `InternalPreserveStackTrace` to avoid doing reflection every time
- squashed the fixup commits

adamralph commented on src/FakeItEasy/Core/WrappedObjectRule.cs:
```diff
@@ -50,8 +52,16 @@ public void Apply(IInterceptedFakeObjectCall fakeObjectCall)
+            try
+            {
+                var valueFromWrappedInstance = fakeObjectCall.Method.Invoke(this.wrappedObject, parameters);
+                fakeObjectCall.SetReturnValue(valueFromWrappedInstance);
```
Can this line also throw a `TargetInvocationException`?

thomaslevesque replied:
Good question. I don't think so, but it couldn't hurt to move it outside the `try`. I'll do it

blairconrad commented:
Nice "fixup!".

adamralph commented on tests/FakeItEasy.Specs/WrappingFakeSpecs.cs:
```diff
@@ -0,0 +1,178 @@
+            "And a fake wrapping this object"
+                .x(() => wrapper = A.Fake<IFoo>(o => o.Wrapping(realObject)));
+
+            "And a refParameter variable with a value of 1"
```
nitpick: it's called "ref parameter" elsewhere in the natural language.

adamralph commented on tests/FakeItEasy.Specs/WrappingFakeSpecs.cs:
```diff
@@ -0,0 +1,178 @@
+            Foo realObject,
+            IFoo wrapper,
+            int refParameter,
+            int outParameter)
```
nitpick: the variables referred to as `refParameter` and `outParameter" are really just variables. The method has "out and ref parameters" and the variables are arguments for the method call. Perhaps `refVariable` and `outVariable` would be better names? (Although they still sound a bit strange.)

thomaslevesque replied:
Yes, I could rename them. In fact, I had initially named them `x` and `y`, but FxCop complained that these names weren't meaningful enough...

adamralph replied:
You could even call them `@ref` and `@out`.

thomaslevesque commented:
I made the changes in the specs WRT to the ref/out params, and squashed the fixup commits

blairconrad commented:
Sorry for the delay, @thomaslevesque. It looks very good to me. Thanks!

adamralph commented on tests/FakeItEasy.Specs/WrappingFakeSpecs.cs:
```diff
@@ -0,0 +1,178 @@
+            "And the value of @ref is incremented"
+                .x(() => @ref.Should().Be(2));
+
+            "And the value of @out is set"
```
Sorry to keep nitpicking but I guess this means you intended to rename `outParameter` to `@out`?

blairconrad replied:
Man. I'm going to stop saying how great everything is.

thomaslevesque replied:
> Sorry to keep nitpicking but I guess this means you intended to rename outParameter to @out?

yes ! I thought I had, but apparently not... I will amend (again)

thomaslevesque replied:
Done

adamralph replied:
:ok_hand:

adamralph commented:
Thanks @thomaslevesque!

thomaslevesque commented:
Thanks for the review, and for the merge!
